pkg/sql/resolver.go

```diff
@@ -370,32 +373,39 @@ func expandIndexName(
 		// subsequent call to expandIndexName() will generate tn using the
 		// new value of index.Table, which is a table name.
 
-		// Just an assertion: if we got there, there cannot be a path prefix
-		// in tn or a value in index.Index yet.
-		if tn.ExplicitSchema || tn.ExplicitCatalog || index.Index != "" {
+		// Just an assertion: if we got there, there cannot be a value in index.Index yet.
+		if index.Index != "" {
 			return nil, nil, pgerror.NewErrorf(pgerror.CodeInternalError,
 				"programmer error: not-searched index name found already qualified: %s@%s", tn, index.Index)
 		}
 
-		curDb := sc.CurrentDatabase()
-		if curDb == "" {
-			return nil, nil, pgerror.NewErrorf(pgerror.CodeUndefinedObjectError,
-				"no database specified: %q", tree.ErrString(index))
+		index.Index = tree.UnrestrictedName(tn.TableName)
+
+		// Look up the table prefix.
+		found, _, err := tn.TableNamePrefix.Resolve(ctx, sc, sc.CurrentDatabase(), sc.CurrentSearchPath())
+		if err != nil {
+			return nil, nil, err
+		}
+		if !found {
+			if requireTable {
+				return nil, nil, pgerror.NewErrorf(pgerror.CodeUndefinedObjectError,
+					"no schema has been selected to search index: %q",
+					tree.ErrString(&index.Index)).SetHintf(
+					"check the current database and search_path are valid")
+			}
+			return nil, nil, nil
 		}
 
-		index.Index = tree.UnrestrictedName(tn.TableName)
 		lookupFlags := sc.CommonLookupFlags(ctx, requireTable)
-		tn, desc, err = findTableContainingIndex(
-			sc.LogicalSchemaAccessor(), curDb, index.Index, lookupFlags)
+		var foundTn *tree.TableName
+		foundTn, desc, err = findTableContainingIndex(
+			sc.LogicalSchemaAccessor(), tn.Catalog(), tn.Schema(), index.Index, lookupFlags)
 		if err != nil {
 			return nil, nil, err
-		} else if tn == nil {
-			// NB: tn is nil here if and only if requireTable is
-			// false, otherwise err would be non-nil.
-			return nil, nil, nil
+		} else if foundTn != nil {
+			// Memoize the table name that was found.
```

Comment:
Is this comment still true? Doesn't look like we're sticking anything onto the struct anymore, unless `tn` points into the struct as a result of having been returned from `Normalize`, which I think would merit a comment.

Reply:
Yes it is still true. I added a comment.